examples/platform/efr32/heap_4_silabs.c

```diff
@@ -0,0 +1,610 @@
+                        pxBlock->xBlockSize        = xWantedSize;
+
+                        /* Insert the new block into the list of free blocks. */
+                        prvInsertBlockIntoFreeList(pxNewBlockLink);
+                    }
+                    else
+                    {
+                        mtCOVERAGE_TEST_MARKER();
+                    }
+
+                    xFreeBytesRemaining -= pxBlock->xBlockSize;
+
+                    if (xFreeBytesRemaining < xMinimumEverFreeBytesRemaining)
+                    {
+                        xMinimumEverFreeBytesRemaining = xFreeBytesRemaining;
+                    }
+                    else
+                    {
+                        mtCOVERAGE_TEST_MARKER();
+                    }
+
+                    /* The block is being returned - it is allocated and owned
+                    by the application and has no "next" block. */
+                    pxBlock->xBlockSize |= xBlockAllocatedBit;
+                    pxBlock->pxNextFreeBlock = NULL;
+                    xNumberOfSuccessfulAllocations++;
+                }
+                else
+                {
+                    mtCOVERAGE_TEST_MARKER();
+                }
+            }
+            else
+            {
+                mtCOVERAGE_TEST_MARKER();
+            }
+        }
+        else
+        {
+            mtCOVERAGE_TEST_MARKER();
+        }
+
+        traceMALLOC(pvReturn, xWantedSize);
+    }
+    (void) xTaskResumeAll();
+
+#if (configUSE_MALLOC_FAILED_HOOK == 1)
+    {
+        if (pvReturn == NULL)
+        {
+            extern void vApplicationMallocFailedHook(void);
+            vApplicationMallocFailedHook();
+        }
+        else
+        {
+            mtCOVERAGE_TEST_MARKER();
+        }
+    }
+#endif
+
+    configASSERT((((size_t) pvReturn) & (size_t) portBYTE_ALIGNMENT_MASK) == 0);
+    return pvReturn;
+}
+/*-----------------------------------------------------------*/
+
+void * pvPortCalloc(size_t num, size_t size)
+{
+    void * ptr = pvPortMalloc(num * size);
+    if (ptr != NULL)
+    {
+        memset(ptr, 0, num * size);
+    }
+
+    return ptr;
+}
+
+void * pvPortRealloc(void * pv, size_t size)
+{
+    void * resized_ptr  = NULL;
+    size_t current_size = xPortPointerSize(pv);
+
+    if (current_size > 0) // pv is allocated
+    {
+        if (size) // New size is not 0
+        {
+
+            if (size == current_size) // if existing pointer is the same size
+            {
+                resized_ptr = pv;
+            }
+            else // New size is a different from current size
+            {
+                resized_ptr = pvPortCalloc(1, size);
+                if (resized_ptr != NULL)
+                {
+                    int smallest_size = size < current_size ? size : current_size;
+                    memcpy(resized_ptr, pv, smallest_size);
+                }
+                vPortFree(pv);
```

Comment:
Shouldn't this `vPortFree` be inside the `if (resized_ptr != NULL)` block?  The usual semantics for realloc is that if it fails the input pointer remains valid.

Reply:
Yes it should be, I pushed the change.